libpimeval/src/pimUtils.h

```diff
@@ -15,18 +15,65 @@
 #include <atomic>
 #include <mutex>
 #include <condition_variable>
-
+#include <algorithm>
+#include <cctype>
+#include <locale>
+#include <unordered_map>
 
 namespace pimUtils
 {
   std::string pimStatusEnumToStr(PimStatus status);
   std::string pimDeviceEnumToStr(PimDeviceEnum deviceType);
+  PimDeviceEnum strToPimDeviceEnum(const std::string& deviceTypeStr);
   std::string pimAllocEnumToStr(PimAllocEnum allocType);
   std::string pimCopyEnumToStr(PimCopyEnum copyType);
   std::string pimDataTypeEnumToStr(PimDataType dataType);
+  std::string& ltrim(std::string& s);
+  std::string& rtrim(std::string& s);
+  std::string& trim(std::string& s);
+  bool readFileContent(const char* fileName, std::string& fileContent);
+  std::string getParam(const std::unordered_map<std::string, std::string>& params, const std::string& key);
+  std::string getOptionalParam(const std::unordered_map<std::string, std::string>& params, const std::string& key, bool& returnStatus);
+  std::string removeAfterSemicolon(const std::string &input);
 
   std::vector<bool> readBitsFromHost(void* src, uint64_t numElements, unsigned bitsPerElement);
   bool writeBitsToHost(void* dest, const std::vector<bool>& bits);
+  std::string getDirectoryPath(const std::string& filePath);
+  bool getEnvVar(const std::string &varName, std::string &varValue);
+
+  const std::unordered_map<PimDeviceEnum, std::string> enumToStrMap = {
+      {PIM_DEVICE_NONE, "PIM_DEVICE_NONE"},
+      {PIM_FUNCTIONAL, "PIM_FUNCTIONAL"},
+      {PIM_DEVICE_BITSIMD_V, "PIM_DEVICE_BITSIMD_V"},
+      {PIM_DEVICE_BITSIMD_V_NAND, "PIM_DEVICE_BITSIMD_V_NAND"},
+      {PIM_DEVICE_BITSIMD_V_MAJ, "PIM_DEVICE_BITSIMD_V_MAJ"},
+      {PIM_DEVICE_BITSIMD_V_AP, "PIM_DEVICE_BITSIMD_V_AP"},
+      {PIM_DEVICE_DRISA_NOR, "PIM_DEVICE_DRISA_NOR"},
+      {PIM_DEVICE_DRISA_MIXED, "PIM_DEVICE_DRISA_MIXED"},
+      {PIM_DEVICE_SIMDRAM, "PIM_DEVICE_SIMDRAM"},
+      {PIM_DEVICE_BITSIMD_H, "PIM_DEVICE_BITSIMD_H"},
+      {PIM_DEVICE_FULCRUM, "PIM_DEVICE_FULCRUM"},
+      {PIM_DEVICE_BANK_LEVEL, "PIM_DEVICE_BANK_LEVEL"}
+  };
+
+  const std::unordered_map<std::string, PimDeviceEnum> strToEnumMap = {
```

Comment:
It's better to keep just one unordered_map from enum to string. It's okay to be O(n) when converting from string to enum, since it is only used for config parsing.

Reply:
I actually think @mhgholamrezaei's approach is cleaner and we should keep it.

Reply:
Okay, this is optional.